atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java

```diff
@@ -166,26 +169,50 @@ public TransactionAndImmutableTsLock setupRunTaskWithConditionThrowOnConflict(Pr
         StartIdentifiedAtlasDbTransactionResponse transactionResponse
                 = timelockService.startIdentifiedAtlasDbTransaction();
         try {
-            LockToken immutableTsLock = transactionResponse.immutableTimestamp().getLock();
-            long immutableTs = transactionResponse.immutableTimestamp().getImmutableTimestamp();
-            recordImmutableTimestamp(immutableTs);
-
-            cleaner.punch(transactionResponse.startTimestampAndPartition().timestamp());
-            Supplier<Long> startTimestampSupplier = Suppliers.ofInstance(
-                    transactionResponse.startTimestampAndPartition().timestamp());
-
-            Transaction transaction = createTransaction(
-                    immutableTs,
-                    startTimestampSupplier,
-                    immutableTsLock,
-                    condition);
-            return TransactionAndImmutableTsLock.of(transaction, immutableTsLock);
+            return wrapResponse(condition, transactionResponse);
         } catch (Throwable e) {
             timelockService.tryUnlock(ImmutableSet.of(transactionResponse.immutableTimestamp().getLock()));
             throw Throwables.rewrapAndThrowUncheckedException(e);
         }
     }
 
+    @Override
+    public List<TransactionAndImmutableTsLock> setupRunTaskBatchWithConditionThrowOnConflict(
+            List<PreCommitCondition> conditions) {
+        Preconditions.checkArgument(!conditions.isEmpty(), "Trying to start an empty batch");
+        try (StartIdentifiedAtlasDbTransactionResponseBatch responses =
+                timelockService.startIdentifiedAtlasDbTransactionsBatch(conditions.size())) {
+            if (responses.size() != conditions.size()) {
+                throw new TransactionBatchFailedRetriableException(
+                        "The number of transactions started does not match the size of the batch.");
+            }
+
+            List<TransactionAndImmutableTsLock> transactions = Streams.zip(
+                    responses.getResponses().stream(),
+                    conditions.stream(),
+                    (response, condition) -> wrapResponse(condition, response)).collect(Collectors.toList());
+            return responses.successful(transactions);
+        }
+    }
+
+    private TransactionAndImmutableTsLock wrapResponse(PreCommitCondition condition,
+            StartIdentifiedAtlasDbTransactionResponse transactionResponse) {
+        LockToken immutableTsLock = transactionResponse.immutableTimestamp().getLock();
+        long immutableTs = transactionResponse.immutableTimestamp().getImmutableTimestamp();
+        recordImmutableTimestamp(immutableTs);
```

Comment:
We can probably optimise this in the batched case (only update once), if that is a performance concern.

Reply:
@jeremyk-91 to help verify; also probably need metrics to help answer this